blog/2024-02-23-transforming-your-data-2/index.md

```diff
@@ -0,0 +1,446 @@
+Jennifer and Cynthia from the previous query each reported disputed transactions in _different_
+merchant locations. However, they all had a common merchant in Boston (ID 70) belonging to Walmart, that
+they all made transactions within a plausible time window. This could indicate that this particular
+merchant, located in Boston, could be a source of fraud.
+
+## Graph algorithms
+
+Kùzu is well-integrated with the PyData ecosystem, including PyTorch Geometric, Pandas, and
+[NetworkX](https://networkx.org/documentation/stable/reference/index.html), a popular Python library
+for network analysis. Because Kùzu is an embedded graph database, it runs in-process with a Python
+application, so it's simple to isolate a subgraph of interest via Cypher and convert it
+to a NetworkX directed graph (DiGraph) for further analysis.
+
+### Weakly connected components
+
+A weakly connected component is a maximal subgraph in which there is a path between any two nodes,
+ignoring the direction of the edges. In the context of our subgraph of clients, merchants and the
+vicinity of disputed transactions, we can use this algorithm to isolate the disjoint subgraphs that
+we saw in the visualization earlier.
+
+We first isolate the subgraph of clients, merchants and transactions in the vicinity of disputed
+transactions as follows:
+
+```py
+import kuzu
+import networkx as nx
+
+db = kuzu.Database("./transaction_db")
+conn = kuzu.Connection(db)
+
+# Isolate the subgraph of interest
+disputed_vicinity = conn.execute(
+    """
+    MATCH (c1:Client)-[t1:TransactedWith]->(m:Merchant)<-[t2:TransactedWith]-(c2:Client)
+    WHERE t1.is_disputed = true
+    RETURN *;
+    """
+)
+
+# Convert to networkx DiGraph
+G1 = clients.get_as_networkx(directed=True)
+```
+
+For every client that reported a disputed transaction, we isolate the subgraph of all the clients
+who interacted with these same merchants, as well as others. Running the weakly connected components
+algorithm will first inform us whether the vicinity of disputed transactions is a single connected
+component or not.
+
+```py
+num_weakly_connected_components = nx.number_weakly_connected_components(G1)
+print(num_weakly_connected_components)
+
+# Output
+2
+```
+
+We obtain 2 weakly connected components, indicating that there are disjoint components in the
+disputed transactions subgraph.
+
+To see which clients/merchants are in each weakly connected component, we can run the following:
+
+```py
+weakly_connected_components = list(nx.weakly_connected_components(G2))
+
+print(len(weakly_connected_components[0]))
+print(len(weakly_connected_components[1]))
+
+# Output
+52
+9
+```
+
+The first cluster of clients and merchants contains 52 nodes, while the second cluster contains 9
+nodes. This helps narrow down on a transaction subgraph that has a greater degree of connectivity.
+
+### Closeness centrality
+
+Centrality algorithms are among the most commonly used algorithms in graph data science. They are
```

Comment:
I would give a high level motivation here. For example: "Centrality algorithms are among the most commonly used algorithms in graph data science. At a high-level, centrality algorithms are used to identify "important" nodes. Specifically nodes that serve the role of connecting many other nodes in the graph. Suppose the owner of our transaction graph is an e-commerce company. Suppose further that the company wants to promote its "important merchants" to attract more customers to its network. Centrality metrics, similar to the popular PageRank metric, are concrete ways to order nodes in terms of "importance". 

I would also change the "Closeness centrality[^2] is different from common algorithms like PageRank or degree centrality in that they indicate how _close_ a node is to all other nodes." sentence to just say "Closeness centrality[^2] measures how _close_ a node n is to all other nodes by calculating the average of the shortest path length from n to every other node in the graph."